types/redux-define/index.d.ts

```diff
@@ -0,0 +1,82 @@
+// Type definitions for redux-define 1.1
+// Project: https://github.com/smeijer/redux-define
+// Definitions by: Joe Barnett <https://github.com/ensconced>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.1
+
+export as namespace ReduxDefine;
+
+export type SubActionProps<
+    SubAction extends string, // This is necessary to force narrowing of the subaction type, as described here https://github.com/microsoft/TypeScript/issues/30680
+    SubActions extends SubAction[],
+    Namespace extends string | undefined,
+    OwnAction extends string,
+> = {
+    [k in SubActions[number]]: `${Namespace extends string ? `${Namespace}/` : ''}${OwnAction}_${k}`;
+};
+
+export type NamespaceString<Namespace extends Action | string> = Namespace extends Action
+    ? Namespace['ACTION']
+    : Namespace;
+
+export type WithNamespace<OwnAction extends string, Namespace extends string | undefined> = Namespace extends string
+    ? `${Namespace}/${OwnAction}`
+    : OwnAction;
+
+export interface PlainAction<OwnAction extends string, Namespace extends string | undefined> {
+    ACTION: WithNamespace<OwnAction, Namespace>;
+    defineAction: typeof defineActionMethod;
+    toString: () => WithNamespace<OwnAction, Namespace>;
+}
+
+export type Action<
+    OwnAction extends string = string,
+    SubAction extends string = string,
+    SubActions extends SubAction[] | undefined = undefined,
+    Namespace extends string | undefined = undefined,
+> = SubActions extends string[]
+    ? PlainAction<OwnAction, Namespace> & SubActionProps<SubAction, SubActions, Namespace, OwnAction>
+    : PlainAction<OwnAction, Namespace>;
+
+export function defineAction<OwnAction extends string>(actionType: OwnAction): Action<OwnAction>;
+export function defineAction<OwnAction extends string, SubAction extends string, SubActions extends SubAction[]>(
+    actionType: OwnAction,
+    subactions: SubActions,
+): Action<OwnAction, SubAction, SubActions>;
+export function defineAction<
+    OwnAction extends string,
+    SubAction extends string,
+    SubActions extends SubAction[],
+    Namespace extends string | Action,
+>(
+    actionType: OwnAction,
+    subactions: SubActions,
+    namespace: Namespace,
+): Action<OwnAction, SubAction, SubActions, NamespaceString<Namespace>>;
+export function defineAction<OwnAction extends string, Namespace extends string | Action>(
+    actionType: OwnAction,
+    namespace: Namespace,
+): Action<OwnAction, string, undefined, NamespaceString<Namespace>>;
+
+export function defineActionMethod<Parent extends Action, OwnAction extends string>(
```

Comment:
The underlying module doesn't appear to export a function called `defineActionMethod`.  If this is just a helper, consider dropping the `export`.

Reply:
The tests didn't allow removing the `export` directly, that gives this error:

```
ll declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity.
```

But I have rewritten this using arrow function types which are intersected to give the overload.